File: bindings/typescript/scripts/bench-fingerprint-followups.mjs
#!/usr/bin/env node
/**
 * Opt-in proof for the deferred fingerprint work.
 *
 *   npm run build --prefix bindings/typescript
 *   node --expose-gc bindings/typescript/scripts/bench-fingerprint-followups.mjs
 *   bun bindings/typescript/scripts/bench-fingerprint-followups.mjs
 *
 * Set ZIGFITSIO_BENCH_SAMPLES to change the default 15 measured pairs.
 */
import assert from "node:assert/strict";

import * as zf from "../dist/index.js";
import * as ll from "../dist/lowlevel/index.js";
import { colFp as coreColFp } from "../dist/table.js";

const samples = Number(process.env.ZIGFITSIO_BENCH_SAMPLES ?? 15);
if (!Number.isInteger(samples) || samples < 3) throw new Error("ZIGFITSIO_BENCH_SAMPLES must be an integer >= 3");

const encoder = new TextEncoder();
const FNV_OFFSET = 0xcbf29ce484222325n;
const FNV_PRIME = 0x100000001b3n;
const U64_MASK = 0xffffffffffffffffn;
let sink = 0n;

function bytesOf(view) {
  return new Uint8Array(view.buffer, view.byteOffset, view.byteLength);
}

function fnv1a64(bytes) {
  let h = FNV_OFFSET;
  for (const byte of bytes) {
    h ^= BigInt(byte);
    h = (h * FNV_PRIME) & U64_MASK;
  }
  return h;
}

const mix = (h, value) => ((h ^ (value & U64_MASK)) * FNV_PRIME) & U64_MASK;

function legacyColFp(column) {
  if (column.kind === "string") return fnv1a64(encoder.encode(column.values.join("\0")));
  let h = FNV_OFFSET;
  for (const cell of column.values) {
    h = mix(h, BigInt(cell.length));
    h = mix(h, fnv1a64(bytesOf(cell)));
  }
  return h;
}

function percentile(values, fraction) {
  const sorted = [...values].sort((a, b) => a - b);
  return sorted[Math.ceil(sorted.length * fraction) - 1];
}

function timed(fn) {
  const start = process.hrtime.bigint();
  sink ^= fn();
  return Number(process.hrtime.bigint() - start) / 1e6;
}

function compare(name, column) {
  for (let i = 0; i < 3; i++) {
    sink ^= legacyColFp(column);
    sink ^= coreColFp(column);
  }

  const legacy = [];
  const core = [];
  for (let i = 0; i < samples; i++) {
    if ((i & 1) === 0) {
      legacy.push(timed(() => legacyColFp(column)));
      core.push(timed(() => coreColFp(column)));
    } else {
      core.push(timed(() => coreColFp(column)));
      legacy.push(timed(() => legacyColFp(column)));
    }
  }

  const legacyMedian = percentile(legacy, 0.5);
  const coreMedian = percentile(core, 0.5);
  return {
    workload: name,
    legacyMedianMs: legacyMedian.toFixed(3),
    legacyP95Ms: percentile(legacy, 0.95).toFixed(3),
    coreMedianMs: coreMedian.toFixed(3),
    coreP95Ms: percentile(core, 0.95).toFixed(3),
    speedup: `${(legacyMedian / coreMedian).toFixed(2)}x`,
  };
}

function verifyFraming() {
  const split = { kind: "vla", dtype: "i4", repeat: 1, values: [Int32Array.of(1), Int32Array.of(2)] };
  const joined = { kind: "vla", dtype: "i4", repeat: 1, values: [Int32Array.of(1, 2), new Int32Array()] };
  assert.notEqual(legacyColFp(split), legacyColFp(joined));
  assert.notEqual(coreColFp(split), coreColFp(joined));

  const strings = { kind: "string", dtype: "u1", repeat: 5, values: ["alpha", "beta"] };
  const changed = { kind: "string", dtype: "u1", repeat: 5, values: ["alpha", "Beta"] };
  assert.notEqual(coreColFp(strings), coreColFp(changed));
}

function measurePristine(label, hdul, expectedCalls) {
  const original = ll.native.fn.zf_fingerprint128_v1;
  if (typeof original !== "function") throw new Error("fingerprint ABI missing; run the TypeScript build first");
  let calls = 0;
  let bytes = 0;
  let fingerprintNs = 0n;
  ll.native.fn.zf_fingerprint128_v1 = (...args) => {
    calls++;
    bytes += Number(args[1]);
    const start = process.hrtime.bigint();
    try {
      return original(...args);
    } finally {
      fingerprintNs += process.hrtime.bigint() - start;
    }
  };

  try {
    for (let i = 0; i < 3; i++) sink ^= BigInt(hdul.toBytes().byteLength);
    const fingerprint = [];
    const total = [];
    for (let i = 0; i < samples; i++) {
      calls = 0;
      bytes = 0;
      fingerprintNs = 0n;
      const start = process.hrtime.bigint();
      const output = hdul.toBytes();
      total.push(Number(process.hrtime.bigint() - start) / 1e6);
      fingerprint.push(Number(fingerprintNs) / 1e6);
      sink ^= BigInt(output.byteLength);
      assert.equal(calls, expectedCalls);
      assert.equal(bytes, 8 * 1024 * 1024);
    }
    return {
      workload: label,
      calls: expectedCalls,
      hashedMiB: (bytes / 1048576).toFixed(1),
      fingerprintMedianMs: percentile(fingerprint, 0.5).toFixed(3),
      fingerprintP95Ms: percentile(fingerprint, 0.95).toFixed(3),
      totalMedianMs: percentile(total, 0.5).toFixed(3),
      totalP95Ms: percentile(total, 0.95).toFixed(3),
    };
  } finally {
    ll.native.fn.zf_fingerprint128_v1 = original;
    hdul.close();
  }
}

function imageFixture() {
  const pixels = new Int16Array(4 * 1024 * 1024);
  pixels[pixels.length - 1] = 1;
  const source = new zf.HDUList([new zf.PrimaryHDU({ data: new zf.FitsArray(pixels, [2048, 2048]) })]).toBytes();
  const hdul = zf.fromBytes(source, "update");
  void hdul.get(0).data;
  return hdul;
}

function tableFixture() {
  const columns = [];
  for (let i = 0; i < 8; i++) {
    const values = new Int32Array(256 * 1024);
    values[values.length - 1] = i + 1;
    columns.push(new zf.Column(`C${i}`, "1J", { array: values }));
  }
  const source = new zf.HDUList([new zf.PrimaryHDU(), zf.BinTableHDU.fromColumns(columns)]).toBytes();
  const hdul = zf.fromBytes(source, "update");
  void hdul.get(1).data;
  return hdul;
}

verifyFraming();

const shortStrings = Array.from({ length: 100_000 }, (_, i) => (i % 4 === 0 ? "" : `s${String(i).padStart(7, "0")}`));
const largeStrings = Array.from({ length: 8192 }, () => "x".repeat(1024));
const tinyVlas = Array.from({ length: 100_000 }, (_, i) => new Int32Array(i % 8));
const largeVlas = Array.from({ length: 8 }, () => new Uint8Array(1024 * 1024));
const thresholdCell = new Uint8Array(1024 * 1024);
const thresholdVlas = Array.from({ length: 65 }, () => thresholdCell);

const runtime = process.versions.bun ? `bun ${process.versions.bun}` : `${process.release.name} ${process.version}`;
console.log(`fingerprint follow-up proof (${samples} alternating samples, ${runtime})`);
console.table([
  compare("strings: 100k short/empty", { kind: "string", dtype: "u1", repeat: 8, values: shortStrings }),
  compare("strings: 8k x 1KiB ASCII", { kind: "string", dtype: "u1", repeat: 1024, values: largeStrings }),
  compare("VLA: 100k cells, lengths 0..7", { kind: "vla", dtype: "i4", repeat: 1, values: tinyVlas }),
  compare("VLA: 8 x 1MiB cells", { kind: "vla", dtype: "u1", repeat: 1, values: largeVlas }),
  compare("VLA: 65 x 1MiB cells (streamed)", { kind: "vla", dtype: "u1", repeat: 1, values: thresholdVlas }),
]);

console.log("pristine update-mode toBytes fingerprint traffic");
console.table([
  measurePristine("8 MiB image", imageFixture(), 1),
  measurePristine("8 x 1 MiB table columns", tableFixture(), 8),
]);

// Observable use of every digest, without making timing results into assertions.
if (sink === -1n) console.log("");

File: bindings/python/scripts/bench_fingerprint_followups.py
#!/usr/bin/env python3
"""Opt-in proof for Python's core-owned fingerprints.

    zig build capi -Doptimize=ReleaseFast
    PYTHONPATH=bindings/python/src uv run --no-project --with numpy \
        python bindings/python/scripts/bench_fingerprint_followups.py

Set ZIGFITSIO_BENCH_SAMPLES to change the default 15 measured pairs.
"""

from __future__ import annotations

from hashlib import blake2b
import math
import os
from pathlib import Path
import platform
from statistics import median
import sys
from time import perf_counter_ns

import numpy as np

if sys.platform == "darwin":
    _subdir, _libname = "lib", "libzigfitsio_capi.dylib"
elif sys.platform == "win32":
    _subdir, _libname = "bin", "zigfitsio_capi.dll"
else:
    _subdir, _libname = "lib", "libzigfitsio_capi.so"
_built_library = Path(__file__).resolve().parents[3] / "zig-out" / _subdir / _libname
if not _built_library.exists():
    raise SystemExit("run `zig build capi -Doptimize=ReleaseFast` before this benchmark")
os.environ["ZIGFITSIO_LIBRARY"] = str(_built_library)

from zigfitsio.core import _col_fp, _ndarray_fp


SAMPLES = int(os.environ.get("ZIGFITSIO_BENCH_SAMPLES", "15"))
if SAMPLES < 3:
    raise ValueError("ZIGFITSIO_BENCH_SAMPLES must be an integer >= 3")

_sink = 0


def _legacy_hash_array_into(digest, arr) -> None:
    arr = np.asarray(arr)
    if not arr.nbytes:
        return
    if arr.flags.c_contiguous:
        digest.update(memoryview(arr).cast("B"))
        return
    chunks = np.nditer(
        arr,
        flags=["external_loop", "buffered", "zerosize_ok"],
        op_flags=[["readonly", "contig"]],
        order="C",
        buffersize=max(1, (64 * 1024) // max(arr.dtype.itemsize, 1)),
    )
    for chunk in chunks:
        digest.update(memoryview(chunk).cast("B"))


def _legacy_ndarray_fp(arr) -> bytes:
    digest = blake2b(digest_size=16)
    _legacy_hash_array_into(digest, arr)
    return digest.digest()


def _legacy_col_fp(col) -> bytes:
    if col.dtype == object:
        digest = blake2b(digest_size=16)
        for cell in col:
            arr = np.asarray(cell)
            digest.update(int(arr.nbytes).to_bytes(8, "little"))
            _legacy_hash_array_into(digest, arr)
        return digest.digest()
    return _legacy_ndarray_fp(col)


def _percentile(values: list[float], fraction: float) -> float:
    ordered = sorted(values)
    return ordered[math.ceil(len(ordered) * fraction) - 1]


def _timed(fn) -> float:
    global _sink
    start = perf_counter_ns()
    digest = fn()
    elapsed_ms = (perf_counter_ns() - start) / 1e6
    _sink ^= int.from_bytes(digest[:8], "little")
    return elapsed_ms


def _compare(name: str, logical_bytes: int, legacy, core) -> tuple[str, ...]:
    global _sink
    for _ in range(3):
        for fn in (legacy, core):
            digest = fn()
            _sink ^= int.from_bytes(digest[:8], "little")

    legacy_times: list[float] = []
    core_times: list[float] = []
    for i in range(SAMPLES):
        if i % 2 == 0:
            first, second = (legacy, legacy_times), (core, core_times)
        else:
            first, second = (core, core_times), (legacy, legacy_times)
        first[1].append(_timed(first[0]))
        second[1].append(_timed(second[0]))

    legacy_median = median(legacy_times)
    core_median = median(core_times)
    return (
        name,
        f"{logical_bytes / 1048576:.2f}",
        f"{legacy_median:.3f}",
        f"{_percentile(legacy_times, 0.95):.3f}",
        f"{core_median:.3f}",
        f"{_percentile(core_times, 0.95):.3f}",
        f"{legacy_median / core_median:.2f}x",
    )


def _vla_fixture() -> np.ndarray:
    lengths = (0, 1, 16, 256, 4096)
    cells = np.empty(2048, dtype=object)
    for i in range(cells.size):
        length = lengths[i % len(lengths)]
        base = np.arange(length * 2, dtype="i4")
        cells[i] = base[::2] if i % 4 == 0 else base[:length]
    return cells


def main() -> None:
    contiguous = np.arange(8 * 1024 * 1024, dtype="u1")
    transposed = contiguous.reshape(2048, 4096).T
    vlas = _vla_fixture()

    rows = [
        _compare(
            "8 MiB contiguous u1",
            contiguous.nbytes,
            lambda: _legacy_ndarray_fp(contiguous),
            lambda: _ndarray_fp(contiguous),
        ),
        _compare(
            "8 MiB transposed u1",
            transposed.nbytes,
            lambda: _legacy_ndarray_fp(transposed),
            lambda: _ndarray_fp(transposed),
        ),
        _compare(
            "VLA: 2048 mixed cells",
            sum(8 + np.asarray(cell).nbytes for cell in vlas),
            lambda: _legacy_col_fp(vlas),
            lambda: _col_fp(vlas),
        ),
    ]

    print(
        f"Python fingerprint proof ({SAMPLES} alternating samples, "
        f"Python {platform.python_version()}, NumPy {np.__version__})"
    )
    headers = ("workload", "MiB", "legacy med ms", "legacy p95 ms", "core med ms", "core p95 ms", "speedup")
    widths = [max(len(headers[i]), *(len(row[i]) for row in rows)) for i in range(len(headers))]
    print("  ".join(value.ljust(widths[i]) for i, value in enumerate(headers)))
    print("  ".join("-" * width for width in widths))
    for row in rows:
        print("  ".join(value.ljust(widths[i]) for i, value in enumerate(row)))

    if _sink == -1:
        print("")


if __name__ == "__main__":
    main()
